pkg/sql/sqlstats/insights/integration/integration_test: TestInsightsIntegrationForContention failed

pkg/sql/sqlstats/insights/integration/integration_test.TestInsightsIntegrationForContention [failed](https://mesolite.cluster.engflow.com/invocations/default/8b652886-a592-4622-8597-6b16ec979007?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvc3FsL3NxbHN0YXRzL2luc2lnaHRzL2ludGVncmF0aW9uOmludGVncmF0aW9uX3Rlc3Q=) on release-25.3.0-rc @ [62ef3d92adaa42eedad17b474fcf89556f215ded](https://github.com/cockroachdb/cockroach/commits/62ef3d92adaa42eedad17b474fcf89556f215ded):


```
=== RUN   TestInsightsIntegrationForContention
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestInsightsIntegrationForContention2229803272
    test_log_scope.go:76: use -show-logs to present logs inline
    test_server_shim.go:156: automatically injected an external process virtual cluster under test; see comment at top of test_server_shim.go for details.
    insights_test.go:834: error scanning '&{<nil> 0xc001fca640}': sql: no rows in result set
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestInsightsIntegrationForContention2229803272
--- FAIL: TestInsightsIntegrationForContention (14.54s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-prs
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestInsightsIntegrationForContention.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-53088

--- a/pkg/sql/sqlstats/insights/integration/BUILD.bazel
+++ b/pkg/sql/sqlstats/insights/integration/BUILD.bazel
@@ -26,6 +26,7 @@ go_test(
         "//pkg/util/log",
         "//pkg/util/timeutil",
         "//pkg/util/uuid",
+        "@com_github_cockroachdb_errors//:errors",
         "@com_github_stretchr_testify//require",
     ],
 )

--- a/pkg/sql/sqlstats/insights/integration/insights_test.go
+++ b/pkg/sql/sqlstats/insights/integration/insights_test.go
@@ -38,6 +38,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
 
@@ -803,7 +804,6 @@ func TestInsightsIntegrationForContention(t *testing.T) {
 	// Chan to wait for the txn to complete to avoid checking for insights before the txn is committed.
 	txnDoneChan := make(chan struct{})
 
-	observerConn := sqlutils.MakeSQLRunner(tc.ApplicationLayer(0).SQLConn(t))
 	txConn := sqlutils.MakeSQLRunner(tc.ApplicationLayer(0).SQLConn(t))
 	tx := txConn.Begin(t)
 
@@ -828,10 +828,20 @@ func TestInsightsIntegrationForContention(t *testing.T) {
 	require.NoError(t, errTxn)
 
 	var waitingTxnID uuid.UUID
-	observerConn.QueryRow(t,
-		`SELECT id
+	testutils.SucceedsSoon(t, func() error {
+		r, err := tc.ApplicationLayer(0).SQLConn(t).Query(`SELECT id
          FROM crdb_internal.node_transactions
-         WHERE application_name = 'waiting_txn'`).Scan(&waitingTxnID)
+         WHERE application_name = 'waiting_txn'`)
+		require.NoError(t, err)
+		defer r.Close()
+
+		if !r.Next() {
+			return errors.New("waiting_txn not found")
+		}
+		err = r.Scan(&waitingTxnID)
+		require.NoError(t, err)
+		return nil
+	})
 
 	require.NoError(t, tx.Commit())
 

